google/cloud/ndb/model.py

```diff
@@ -2494,6 +2494,11 @@ def _from_base_type(self, value):
             indicate that the value didn't need to be unwrapped and
             decompressed.
         """
+        # First, check for legacy compressed LocalStructuredProperty values.
+        # See https://github.com/googleapis/python-ndb/issues/359
+        if self._compressed and isinstance(value, ds_entity_module.Entity):
+            return
+
         if self._compressed and not isinstance(value, _CompressedValue):
             if not value.startswith(_ZLIB_COMPRESSION_MARKER):
                 value = zlib.compress(value)
```

Comment:
Why compress here if we're just going to decompress below? Can we just return?

Reply:
And just below here, in that last `if` statement, don't we already know it's a _CompressedValue at that point? Do we need the `if`?

Reply:
I agree that compressing and then immediately decompressing is foolish. Will change that. I do think the last check is needed, because it could happen that we could be dealing with a non-compressed property.

Reply:
Ah, right.

Reply:
In fact, now that I look at it,  I think that check is actually the only needed code in that method. Thanks for bringing it to my attention.

Reply:
Ok, I had to follow the different possibilities step by step to get a better grasp. The only possible improvement was your first suggestion to avoid compressing and return instead. Done.